### ISSUE TITLE
Flutter Widget Inspector does not render correctly, mostly a blank page

### ISSUE THREAD

**Issue by lukehutch** [Dart-Code/Dart-Code#4855]
**Describe the bug**

My Flutter Widget Inspector randomly broke a few weeks ago, and I haven't been able to figure out how to get it working again. It's either blank, or displays a few lines like this:

![image](https://github.com/Dart-Code/Dart-Code/assets/811305/99cd1b0d-435d-46fb-968f-e7c95e5683d4)

<details>

<summary><strong>Workspace Environment</strong></summary>

```text
Dart Code extension: 3.77.20231116
Flutter extension: 3.77.20231101 (activated)

App: Visual Studio Code
App Host: desktop
Version: linux 1.84.2

Workspace type: Dart, Flutter (LSP) (Multiroot)

Dart (3.3.0-91.0.dev): /opt/flutter/bin/cache/dart-sdk
Flutter (3.17.0-0.0.pre): /opt/flutter (sdk gphone x86 64 (android-x64/android))

```

</details>

<details>

<summary><strong>Output from 'dart info'</strong></summary>

`/opt/flutter/bin/cache/dart-sdk/bin/dart info`

If providing this information as part of reporting a bug, please review the information
below to ensure it only contains things you're comfortable posting publicly.

#### General info

- Dart 3.3.0-91.0.dev (dev) (Thu Nov 2 09:10:03 2023 -0700) on "linux_x64"
- on linux / Linux 6.5.10-200.fc38.x86_64 #1 SMP PREEMPT_DYNAMIC Thu Nov  2 19:59:55 UTC 2023
- locale is en_US.utf8

#### Project info

- sdk constraint: '>=3.1.5 <4.0.0'
- dependencies: auto_size_text, boxy, cached_network_image, carousel_slider, collection, cross_file, cupertino_icons, device_info_plus, email_validator, exif, extended_image, flutter, flutter_cache_manager, flutter_email_sender, flutter_fgbg, flutter_image_utilities, flutter_localizations, flutter_persistent_value_notifier, flutter_reactive_value, flutter_svg, fluttertoast, geocoding, geolocator, getwidget, google_fonts, google_mlkit_face_detection, http, image_picker, intl, jiffy, line_awesome_flutter, logging, material_dialogs, package_info_plus, path_provider, persistent_bottom_nav_bar_v2, provider, random_string, routemaster, share_plus, shared_preferences, url_launcher
- dev_dependencies: flutter_launcher_icons, flutter_lints, flutter_test, lints
- elided dependencies: 11

#### Process info

|  Memory |  CPU | Elapsed time | Command line                                                                               |
| ------: | ---: | -----------: | ------------------------------------------------------------------------------------------ |
|  525 MB | 0.0% |   1-04:50:32 | dart --enable-vm-service=0 --pause_isolates_on_start --disable-dart-dev -DSILENT_VM_SERVICE=true --write-service-info=file:<path>/vm.json --pause_isolates_on_exit --enable-asserts <path>/main.dart |
|  714 MB | 0.0% |   1-18:50:10 | dart ..<path>/serverpod_cli.dart generate --watch                                          |
|   77 MB | 0.0% |   1-04:50:32 | dart debug_adapter                                                                         |
|   75 MB | 0.0% |   3-04:10:13 | dart devtools --machine --allow-embedding --port 9100                                      |
|   64 MB | 0.0% |   4-01:44:58 | dart devtools --machine --allow-embedding --port 9101                                      |
| 2174 MB | 0.4% |   3-04:10:13 | dart language-server --protocol=lsp --client-id=VS-Code --client-version=3.77.20231116     |
| 1480 MB | 0.2% |   2-00:32:28 | dart language-server --protocol=lsp --client-id=VS-Code --client-version=3.77.20231116     |
|   94 MB | 0.0% |   4-01:44:58 | flutter_tools.snapshot daemon                                                              |
|   94 MB | 0.0% |   3-04:10:13 | flutter_tools.snapshot daemon                                                              |
|   74 MB | 0.1% |        51:18 | flutter_tools.snapshot debug_adapter                                                       |
|  170 MB | 0.2% |        51:18 | flutter_tools.snapshot run --machine --start-paused -d emulator-5554 --devtools-server-address http:<path>/ --target <path>/main.dart |

</details>

<details>

<summary><strong>Output from 'flutter doctor'</strong></summary>

`/opt/flutter/bin/flutter doctor -v`

```text
[✓] Flutter (Channel beta, 3.17.0-0.0.pre, on Fedora Linux 38 (Workstation Edition) 6.5.10-200.fc38.x86_64, locale en_US.utf8)
    • Flutter version 3.17.0-0.0.pre on channel beta at /opt/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 0b591f2c82 (3 weeks ago), 2023-11-02 22:01:10 -0400
    • Engine revision dff66925dc
    • Dart version 3.3.0 (build 3.3.0-91.0.dev)
    • DevTools version 2.29.0

[✓] Android toolchain - develop for Android devices (Android SDK version 33.0.1)
    • Android SDK at /home/luke/Android/Sdk
    • Platform android-UpsideDownCake, build-tools 33.0.1
    • Java binary at: /opt/android-studio/jbr/bin/java
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-17.0.6b802.4-9586694)
    • All Android licenses accepted.

[✓] Chrome - develop for the web
    • Chrome at google-chrome

[✓] Linux toolchain - develop for Linux desktop
    • clang version 16.0.6 (Fedora 16.0.6-3.fc38)
    • cmake version 3.27.7
    • ninja version 1.11.1
    • pkg-config version 1.8.0

[✓] Android Studio (version 2022.2)
    • Android Studio at /opt/android-studio
    • Flutter plugin version 73.0.2
    • Dart plugin version 222.4582
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-17.0.6b802.4-9586694)

[✓] VS Code (version 1.84.2)
    • VS Code at /usr/share/code
    • Flutter extension version 3.77.20231101

[✓] Connected device (3 available)
    • sdk gphone x86 64 (mobile) • emulator-5554 • android-x64    • Android 13 (API 33) (emulator)
    • Linux (desktop)            • linux         • linux-x64      • Fedora Linux 38 (Workstation Edition) 6.5.10-200.fc38.x86_64
    • Chrome (web)               • chrome        • web-javascript • Google Chrome 119.0.6045.159

[✓] Network resources
    • All expected network resources are available.

• No issues found!
```

</details>

**Comment by lukehutch** [comment]
`flutter clean`, uninstalling and reinstalling the Flutter VS Code extension, restarting VS Code -- nothing seems to solve this.

**Comment by DanTup** [comment]
@lukehutch if you run the **Open DevTools** command and then choose to **Open in Browser**, do you see the same thing or does it only occur when it's embedded inside VS Code?

>  randomly broke a few weeks ago,

I see you're currently on beta 3.17 - are you able to see whether it occurs on current stable 3.16 (if not, that narrows things down a little).

@kenzieschmoll I don't suppose you've seen anything like this before?

**Comment by lukehutch** [comment]
I do have a massive widget tree, so on the web browser view, it takes about 30 seconds to display anything. But waiting even longer than that on VS Code, nothing is shown (or occasionally, a partial render is shown, something like in the screenshot -- it varies).

In other words, maybe there is a memory or time limit that is being hit?

**Comment by DanTup** [comment]
Can you reproduce this on any public projects that I'd be able to do some testing with?

**Comment by lukehutch** [comment]
I can reproduce this with this small reproducer project that I made for this bug: https://github.com/flutter/flutter/issues/136642#issuecomment-1767044044

![image](https://github.com/Dart-Code/Dart-Code/assets/811305/f4c1aa14-4380-4cba-9a85-4464d202373c)

However, I doubt that the issue has anything to do with any particular Flutter project. The issue is probably due to something being corrupted in a config file or an installed package.

Is there any form of debugging I can enable to get you some more info?

I'd be OK giving you remote access to my machine, if you need it to debug this, but hopefully you can come up with some things I can try first.

**Comment by DanTup** [comment]
Thanks - could you confirm a few things:

- Exact OS/version
- Extact VS Code version (the output of `code --status` when you have this project open might be useful too)
- Any errors in the console (Help -> Toggle Developer Tools -> Console) when you see this?
- Does running code with `code --disable-gpu` make any difference? (close all Code windows before running this as I'm not sure if it re-uses existing processes)
- Does resizing the window (either the whole VS Code window or the split pane to resize the frame) make any difference?

Also - do you have any other machines you can test the same thing on? (seeing it on another machine might increase the odds I'll be able to reproduce it, and it'll likely be easier to debug on a machine set up to run everything from source). Thanks!

**Comment by DanTup** [comment]
I thought those lines were part of the inspector, but actually I don't see those lines in my inspector - it looks like they might actually be the UI guidelines from the editor?

Another thing to check - do you see any rendering issues if you open a markdown file and run the **Markdown: Open Preview to Side** command? I'm wondering if this is a VS Code issue more than a DevTools (or Dart-Code) issue.

**Comment by lukehutch** [comment]
Sometimes a light blue bar appears at the bottom of the widget inspector view:

![image](https://github.com/Dart-Code/Dart-Code/assets/811305/93be1813-8e1a-43f9-b002-a31e748dc46e)

**Comment by lukehutch** [comment]
> * Exact OS/version

```console
$ uname -a
Linux fedora 6.5.10-200.fc38.x86_64 #1 SMP PREEMPT_DYNAMIC Thu Nov  2 19:59:55 UTC 2023 x86_64 GNU/Linux
```

> * Extact VS Code version (the output of `code --status` when you have this project open might be useful too)

```console
$ code --version
1.84.2
1a5daa3a0231a0fbba4f14db7ec463cf99d7768e
x64

$ code --status
Version:          Code 1.84.2 (1a5daa3a0231a0fbba4f14db7ec463cf99d7768e, 2023-11-09T10:50:47.800Z)
OS Version:       Linux x64 6.5.10-200.fc38.x86_64
CPUs:             AMD Ryzen 9 3950X 16-Core Processor (32 x 4245)
Memory (System):  125.70GB (83.03GB free)
Load (avg):       1, 2, 2
VM:               0%
Screen Reader:    no
Process Argv:     . --crash-reporter-id 11879ce0-438e-429c-8d0c-4da9fd848f93
GPU Status:       2d_canvas:                              enabled
                  canvas_oop_rasterization:               disabled_off
                  direct_rendering_display_compositor:    disabled_off_ok
                  gpu_compositing:                        enabled
                  multiple_raster_threads:                enabled_on
                  opengl:                                 enabled_on
                  rasterization:                          enabled
                  raw_draw:                               disabled_off_ok
                  video_decode:                           enabled
                  video_encode:                           disabled_software
                  vulkan:                                 disabled_off
                  webgl:                                  enabled
                  webgl2:                                 enabled
                  webgpu:                                 disabled_off

CPU %	Mem MB	   PID	Process
    0	   129	3438540	code main
    0	     0	3438547	   zygote
    0	   257	3438582	     gpu-process
    0	     0	3438617	       broker
    0	     0	3438548	   zygote
    0	     0	3438550	     zygote
    0	   257	3449950	window [4] (Widget Inspector - click (Workspace) - Visual Studio Code)
    0	   129	3451827	window [5] (create_test_profiles.dart - click (Workspace) - Visual Studio Code)
    0	     0	3438587	   utility-network-service
    0	     0	3438678	shared-process
    0	     0	3438720	ptyHost
    0	     0	3452040	     /usr/bin/bash --init-file /usr/share/code/resources/app/out/vs/workbench/contrib/terminal/browser/media/shellIntegration-bash.sh
    0	     0	3832618	     /usr/bin/bash --init-file /usr/share/code/resources/app/out/vs/workbench/contrib/terminal/browser/media/shellIntegration-bash.sh
    0	   257	3451848	extensionHost [5]
    0	     0	3451948	     electron-nodejs (config.js )
    0	   901	3451957	     /opt/flutter/bin/cache/dart-sdk/bin/dart language-server --protocol=lsp --client-id=VS-Code --client-version=3.77.20231116
    0	     0	3451971	     /opt/flutter/bin/cache/dart-sdk/bin/dart devtools --machine --allow-embedding
    0	     0	3452560	     electron-nodejs (/usr/share/code/code --ms-enable-electron-run-as-node /home/luke/.vscode/extensions/redhat.fabric8-analytics-0.7.3/dist/server.js --node-ipc --clientProcessId=3451848)
    0	     0	3451859	fileWatcher [5]
    0	   386	3832504	extensionHost [4]
    0	     0	3832544	     electron-nodejs (config.js )
    0	  1802	3832553	     /opt/flutter/bin/cache/dart-sdk/bin/dart language-server --protocol=lsp --client-id=VS-Code --client-version=3.77.20231116
    0	     0	3832569	     /opt/flutter/bin/cache/dart-sdk/bin/dart devtools --machine --allow-embedding
    0	     0	3833283	     electron-nodejs (/usr/share/code/code --ms-enable-electron-run-as-node /home/luke/.vscode/extensions/redhat.fabric8-analytics-0.7.3/dist/server.js --node-ipc --clientProcessId=3832504)
    0	   129	3833310	     /usr/java/jdk-11/bin/java -Dfile.encoding=utf8 -jar /home/luke/.vscode/extensions/vscjava.vscode-gradle-3.13.4/lib/gradle-language-server.jar
    0	     0	3833454	     /usr/java/jdk-11/bin/java -Dfile.encoding=UTF-8 -classpath /home/luke/.vscode/extensions/vscjava.vscode-gradle-3.13.4/lib/gradle-server.jar com.github.badsyntax.gradle.GradleServer 45735
    0	   772	3833649	     /home/luke/.vscode/extensions/redhat.java-1.24.0-linux-x64/jre/17.0.8.1-linux-x86_64/bin/java --add-modules=ALL-SYSTEM --add-opens java.base/java.util=ALL-UNNAMED --add-opens java.base/java.lang=ALL-UNNAMED --add-opens java.base/sun.nio.fs=ALL-UNNAMED -Declipse.application=org.eclipse.jdt.ls.core.id1 -Dosgi.bundles.defaultStartLevel=4 -Declipse.product=org.eclipse.jdt.ls.core.product -Djava.import.generatesMetadataFilesAtProjectRoot=false -DDetectVMInstallationsJob.disabled=true -Dfile.encoding=utf8 -XX:+UseParallelGC -XX:GCTimeRatio=4 -XX:AdaptiveSizePolicyWeight=90 -Dsun.zip.disableMemoryMapping=true -Xmx1G -Xms100m -Xlog:disable -javaagent:/home/luke/.vscode/extensions/redhat.java-1.24.0-linux-x64/lombok/lombok-1.18.30.jar -XX:+HeapDumpOnOutOfMemoryError -XX:HeapDumpPath=/home/luke/.config/Code/User/workspaceStorage/9a8af0226dd93409e27ac99b624ed091/redhat.java -Daether.dependencyCollector.impl=bf -jar /home/luke/.vscode/extensions/redhat.java-1.24.0-linux-x64/server/plugins/org.eclipse.equinox.launcher_1.6.500.v20230717-2134.jar -configuration /home/luke/.config/Code/User/globalStorage/redhat.java/1.24.0/config_linux -data /home/luke/.config/Code/User/workspaceStorage/9a8af0226dd93409e27ac99b624ed091/redhat.java/jdt_ws --stdio
    0	     0	3833727	     /home/luke/.vscode/extensions/ms-vscode.cpptools-1.18.5-linux-x64/bin/cpptools
    0	     0	3832514	fileWatcher [4]

Workspace Stats: 
|  Window (create_test_profiles.dart - click (Workspace) - Visual Studio Code)
|  Window (Widget Inspector - click (Workspace) - Visual Studio Code)
|    Folder (click_client): 48 files
|      File types: dart(26) md(2) yaml(2) json(1) gitignore(1) lock(1)
|      Conf files:
|    Folder (click_server): 239 files
|      File types: jpg(52) dart(44) yaml(32) tf(13) json(12) sql(8) sh(4)
|                  csv(3) gitignore(2) md(2)
|      Conf files: launch.json(1) dockerfile(1)
|      Launch Configs: dart(2)
|    Folder (serverpod): 2577 files
|      File types: dart(905) yaml(285) md(88) json(88) lock(42) tf(41)
|                  gitignore(28) sql(26) png(18) pgsql(15)
|      Conf files: dockerfile(10) launch.json(6) github-actions(2)
|                  settings.json(1) tasks.json(1) package.json(1)
|                  tsconfig.json(1)
|      Launch Configs: dart(71)
|    Folder (serverpod): 2577 files
|      File types: dart(905) yaml(285) md(88) json(88) lock(42) tf(41)
|                  gitignore(28) sql(26) png(18) pgsql(15)
|      Conf files: dockerfile(10) launch.json(6) github-actions(2)
|                  settings.json(1) tasks.json(1) package.json(1)
|                  tsconfig.json(1)
|      Launch Configs: dart(71)
|    Folder (click_flutter): 6800 files
|      File types: class(578) dex(311) len(276) json(233) jar(192) xml(189)
|                  bin(168) txt(160) flat(122) tab(100)
|      Conf files: launch.json(1) settings.json(1) cmake(1)
|      Launch Configs: dart(3)
```

> * Any errors in the console (Help -> Toggle Developer Tools -> Console) when you see this?

OK, I think you hit the nail on the head... there are thousands of lines of shader compilation errors, like the following, and DevTools is brought to its knees by the amount of output:

![image](https://github.com/Dart-Code/Dart-Code/assets/811305/1ee87062-4956-45f7-a886-7149f6be55e6)

> * Does running code with `code --disable-gpu` make any difference? (close all Code windows before running this as I'm not sure if it re-uses existing processes)

I closed all windows and re-opened with `--disable-gpu`. Now I see only a gray box in the place of the Widget Inspector:

![image](https://github.com/Dart-Code/Dart-Code/assets/811305/02844e76-51d6-4212-a29d-51de3a1e58a3)

These are the new errors in the DevTools console:

![image](https://github.com/Dart-Code/Dart-Code/assets/811305/4ee4b20b-e565-45e3-bf40-ddad2e5b2011)

> * Does resizing the window (either the whole VS Code window or the split pane to resize the frame) make any difference?

No.

> Also - do you have any other machines you can test the same thing on? (seeing it on another machine might increase the odds I'll be able to reproduce it, and it'll likely be easier to debug on a machine set up to run everything from source). Thanks!

I do have another machine, but I'll have to set it up with Flutter etc. It has a different GPU, so I suspect there won't be a problem on that machine. I'll wait on your response to the above before I go to the effort of setting up that machine.

**Comment by lukehutch** [comment]
> I thought those lines were part of the inspector, but actually I don't see those lines in my inspector - it looks like they might actually be the UI guidelines from the editor?

I think they are fragments of the lines in the tree view.

> Another thing to check - do you see any rendering issues if you open a markdown file and run the **Markdown: Open Preview to Side** command? I'm wondering if this is a VS Code issue more than a DevTools (or Dart-Code) issue.

There is no issue with "Markdown: Open Preview to the Side", either with or without the GPU enabled.

**Comment by DanTup** [comment]
Thanks - my guess is that it's related to those issues with canvaskit:

![image](https://github.com/Dart-Code/Dart-Code/assets/1078012/bafb47fc-88f0-42c7-bf56-045a6f96f9dd)

But I'm not sure what combinations of what things are involved here. VS Code is shipping an older version of Chrome (114 from May 30, 2023) which might be a factor, but that can't be the whole story because it works for others. 

These are a bit of a shot in the dark, but might be worth trying if you get chance:

- Are there and GPU driver updates you can install?
- If you install the live [preview extension](https://marketplace.visualstudio.com/items?itemName=ms-vscode.live-server) which gives you a browser inside VS Code, does the issue reproduce if you:
  - run `dart devtools` in a terminal and then paste that URL into the in-Code browser preview
  - run your own Flutter app compiled with canvaskit (not DevTools) from a terminal and paste that in

cc @kenzieschmoll @jacob314 in case they have any ideas. Is there any way to force DevTools to use the HTML renderer (I see the release is built with `--web-renderer canvaskit` but I don't know whether it includes a fallback?).

**Comment by lukehutch** [comment]
> * Are there and GPU driver updates you can install?

I am running the latest NVIDIA binary driver. It's up to date.

> * run `dart devtools` in a terminal and then paste that URL into the in-Code browser preview

That gives me the error "Externally hosted links are not supported in the embedded preview".

> * run your own Flutter app compiled with canvaskit (not DevTools) from a terminal and paste that in

I'm not sure how to do this, but I'm also not sure if I should proceed with figuring this out, given the comments in the other bug you created in flutter/devtools...

**Comment by lukehutch** [comment]
Confirmed, this is fixed with Code Insiders, so I'll go ahead and close this. Thanks for your attention to the issue!

**Comment by DanTup** [comment]
Excellent, thanks for confirming :)

**Comment by lukehutch** [comment]
This is broken again in Code Insiders:

![image](https://github.com/Dart-Code/Dart-Code/assets/811305/5c3baf1d-6361-4710-9324-c8cdf6c8119c)

**Comment by DanTup** [comment]
@lukehutch when do you see this? I just tested in Insiders and although I couldn't repro it being completely broken, I did see it flash this colour (for around a second) when I stopped my debug session, before it rendered the "Disconnected" page:

![image](https://github.com/Dart-Code/Dart-Code/assets/1078012/06659e3b-760c-498c-8d8d-41a679a20054)

Are you seeing it when you stop a session, or as soon as you try to load?


> Can we get some better error reporting on the console when the renderer fails?

It's difficult to answer this. If errors occur, they should be shown in the console. If there are no errors and something just isn't rendering properly because of VS Code issues (or GPU issues, etc.) then it's probably not something we can detect. If we can figure out what's going wrong here, we can look at whether we could better detect those types of issues though.

**Comment by lukehutch** [comment]
It's as soon as the DevTools tab opens, and it just sits there doing nothing.

Your screen is a darker shade of gray, which might be a theme difference, or (I think more likely) you are probably looking at something different.

I actually forgot to check the JS console. I don't see the shader compilation errors there now, so maybe fixing that exposed a different issue.

This is what I see in the DevTools JS console, which is not very helpful, because the JS is minified:

![image](https://github.com/Dart-Code/Dart-Code/assets/811305/c161e655-2756-4b1f-8558-75ac7d7544b5)

How do I catch the JS exception back in the Dart source code for Widget Inspector?

**Comment by DanTup** [comment]
> Your screen is a darker shade of gray, which might be a theme difference, or (I think more likely) you are probably looking at something different.

Sorry I wasn't very clear - I see your shade of grey for a second or two _before_ it then switches to the disconnect screen. Perhaps there's some background colour that's the shade of grey from your screenshot or something.

Can you confirm the exact version of Flutter you're on and whether opening DevTools in the browser (from VS Code) works ok? That exception in the console certainly seems suspicious.

**Comment by lukehutch** [comment]
```console
$ flutter --version
Flutter 3.17.0-1.0.pre.3 • channel beta • https://github.com/flutter/flutter.git
Framework • revision 12b47270b7 (2 weeks ago) • 2023-11-27 17:21:11 -0600
Engine • revision dff66925dc
Tools • Dart 3.3.0 (build 3.3.0-91.0.dev) • DevTools 2.29.0
```

Yes, opening DevTools in the browser does work. (I forgot I could do that... thanks)

**Comment by DanTup** [comment]
@lukehutch is this still happening? I just noticed this issue was open but marked as "Fixed in VS Code".

I suspect if it's still broken that it's probably an upstream issue we can't fix here though.

**Comment by lukehutch** [comment]
Looks like it is fixed in the RedHat RPM release `code-1.86.2-1707854644.el8.x86_64`.

Thanks -- I'll close this!
